### DISCUSSION TITLE
Suggestion of improvements

### DISCUSSION THREAD

**Discussion by emmanuel-ruffio** [binjr/binjr#160] [Ideas]
I have been using Binjr for severals months. It works pretty well.
I have some "minor" suggestions that would improve the software:
1) Ability to filter/smooth data: not necessarily a complex frequency-based filter, but just a moving average and/or median (in case of outliers) with user-specified window width (in samples or in times). 

And possibly, in the long term:
2) Ability to shift data by adding a constant value: it's useful to compare data
3) Ability to select the time window over which statistics are performed (min/max/average)
4) Ability to create new data series by performing math operations over current data series. Ex:    y(new) = data1 - data2
5) Ability to create data hierarchy with csv files based on column names:    group1.data1; group1.data2; group2.data1 etc...

Best regards.

**Comment by fthevenet** [comment]
Hi,

Thanks for your suggestions, it is always very insightful to get user feedback.

The addition of post-processing stage for series that would be configurable - or even programmable - by the end user on a per-series basis would be the ideal way to introduce 1) 2) and 4).

binjr already has a transform function pipeline that series go through before being displayed, that is used to clean-up, sort and align the raw data, as well doing downsampling, but at the moment those are applied to all series and, with the exception of downsampling, offer very little in terms of configuration by the user.

There's quite of bit of work in getting this right though, and a fair amount of problem to solve first: where/how should the new component for this be added in the existing UI? Should it introduce an expression language in order to limit the complexity of these UI components and increase the flexibility offered to the user? Or should it go all the way and add some scripting capabilities? If so, what language? At what point does this turn into an Excel clone? :)

These are all very interesting questions, but ultimately what happens will likely be heavily influenced by what binjr wants to be next (which, I have to say, I've not really figured out yet....). The app has, changed quite a lot over the years - all for the better - and many of its most useful features where born out of outside influences, ideas or suggestions.
I therefore wholeheartedly invite any binjr user out there who is interested in pushing this forward to chime in and participate in that discussion!

Cheers!
